### PR TITLE
[Mentions v2] Support mentions in editing comments

### PR DESCRIPTION
**Pull request by war-in**
### Fixed Issues

$ https://github.com/Expensify/App/issues/39550
PROPOSAL:

### Tests

1. Go to a room
2. Create a comment with user and room mentions
3. Click on `Edit comment`
4. Verify that all mentions are in the edit view
5. Modify comment by adding at least one user and one room mentions (you can remove some mentions too)
6. Save the comment
7. Verify that all previous and new mentions are there
8. Open some `Private notes` and verify that editing mentions works correctly

### QA Steps

1. Go to a room
2. Create a comment with user and room mentions
3. Click on `Edit comment`
4. Verify that all mentions are in the edit view
5. Modify comment by adding at least one user and one room mentions
6. Save the comment
7. Verify that all previous and new mentions are there
8. Open some `Private notes` and verify that editing mentions works correctly

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
        - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/61014013/816c6793-b4c8-4002-8d4a-91f74480b797

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/61014013/3bdfb500-b7f6-498d-af41-618c1f01b9f6

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/61014013/fc525e28-1406-449d-9d91-fcede7f54bad

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/61014013/ecd52ea9-6e51-43d6-95a0-9840c0c3109f

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/61014013/e218d8a6-84a7-4fdd-ae1d-236b6551d409

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/61014013/a5858f46-8e3a-45ca-afff-ddd3737b40a0

</details>

### CODE REVIEW

File: package.json
Comment (war-in) [inline package.json:111]: That will be changed back to `Expensify` once we merge this https://github.com/Expensify/expensify-common/pull/686

File: src/libs/OnyxAwareParser.ts
Comment (war-in) [inline src/libs/OnyxAwareParser.ts:17]: Not sure if we should fallback to id, but we did it before so should be fine

File: src/libs/OnyxAwareParser.ts
Comment (war-in) [inline src/libs/OnyxAwareParser.ts:29]: same here

**Review by robertKozik — APPROVED** [review]
LGTM, lets wait for the `expensify-common` PR merge.
and the task on which we are holding :D

**Comment by abdulrahuman5196** [comment]
I am assuming this PR is on hold. Do let me know otherwise or once it comes to review

**Comment by war-in** [comment]
> I am assuming this PR is on hold. Do let me know otherwise or once it comes to review

@abdulrahuman5196 PR is ready for review :)

**Comment by war-in** [comment]
@abdulrahuman5196 kind bump :)

**Comment by abdulrahuman5196** [comment]
Hi @puneetlath / @war-in I am occupied at the moment. Kindly unassign me and involve a different C+ to move faster.

**Comment by eh2077** [comment]
> Hi @puneetlath / @war-in I am occupied at the moment. Kindly unassign me and involve a different C+ to move faster.

I can help with it.

File: src/libs/OnyxAwareParser.ts
Comment (eh2077) [reply src/libs/OnyxAwareParser.ts:29]: Is it better to fallback to `personalDetails.login` here? On the profile page, we also fallback to show login email if display name is empty
image

**Comment by eh2077** [comment]
@war-in User mention is empty when editing


https://github.com/Expensify/App/assets/117511920/44777a2b-617d-4274-ace9-7ed372674f4c


https://github.com/Expensify/App/assets/117511920/259303c7-8b29-4044-802a-f67fb65402d4

**Comment by war-in** [comment]
@eh2077 thanks for catching it! It seems that we need to bump `expensify-common` dependency and it will be fixed. PR bumping the dependency was reverted recently so I need to take care of it first. Will post an update then

**Comment by war-in** [comment]
Good news :) @blazejkustra is working on the bump ([here](https://github.com/Expensify/App/pull/42387)). Once it's merged our issue should be fixed!

File: src/libs/OnyxAwareParser.ts
Comment (war-in) [reply src/libs/OnyxAwareParser.ts:29]: Actually, we should use the `personalDetails.login` instead of `.displayName` because if we use `.displayName` we won't be able to return from it to the id-based HTML mention tag. Will update it

**Comment by blazejkustra** [comment]
@war-in https://github.com/Expensify/App/pull/42387 was just merged 🚀

**Comment by war-in** [comment]
@eh2077 I've just merged the main and all issues should be fixed 🎉 It's ready for review now :)

**Comment by eh2077** [comment]
I'm working on the checklist

**Review by eh2077 — APPROVED** [review]
LGTM!

File: src/libs/OnyxAwareParser.ts
Comment (war-in) [reply src/libs/OnyxAwareParser.ts:17]: tbh I don't know 😅 It's useful for the users but I'm not sure about the rooms

**Comment by rlinoz** [comment]
I was doing some tests and it looks like we re=render everything that uses `OnyxAwareParser` when we get a new report or a personal detail, I am a little worried about the performance, what do you think @war-in @robertKozik @eh2077?

Am I missing something?

**Comment by war-in** [comment]
@rlinoz That's a good question but I think we have to connect Onyx this way. Do you maybe have an idea of passing arguments which will not cause rerenders?

**Comment by rlinoz** [comment]
Yeah, I played with this a lot today and didn't get to a good solution. Anyway, I believe that a user don't get new reports frequently enough for this to matter, so I think we should be good.

The other option is to fetch only the personal details and reports for a particular message, but that is costly as well and we will have to fetch for ever parsed message, even if we got the same data for a previous one.

**Comment by eh2077** [comment]
> Yeah, I played with this a lot today and didn't get to a good solution. Anyway, I believe that a user don't get new reports frequently enough for this to matter, so I think we should be good.

Yes, I also agree to move the first version of this feature forward.

> The other option is to fetch only the personal details and reports for a particular message, but that is costly as well and we will have to fetch for ever parsed message, even if we got the same data for a previous one.

Just continue to brainstorm on @rlinoz 's idea about improving the potential performance issue. I'm not sure if we have considered just saving the text needed in the `<mention-report>` tag, like
```html
<mention-report reportID="1234" text="#room-name" />
```
So, we won't need to pass the map. But then we'll need to update `<mention-report>` (reportAction) after editing room name.

**Comment by war-in** [comment]
> Just continue to brainstorm on @rlinoz 's idea about improving the potential performance issue. I'm not sure if we have considered just saving the text needed in the <mention-report> tag, like
> ```<mention-report reportID="1234" text="#room-name" />```
> So, we won't need to pass the map. But then we'll need to update <mention-report> (reportAction) after editing room name.

I think the mention with ID was an idea from the design doc. Probably @rlinoz could give more details on it.
However, we would have to change `expensify-common` too :/

**Comment by eh2077** [comment]
> `<mention-report reportID="1234" text="#room-name" />`

As the text can be changed by the user over time, we'll need to check and map it to the correct text by ID. That being said, _**saving the text in the tag doesn’t seem to offer any advantages**_.

**Review by rlinoz — APPROVED** [review]

**Pull request merged**
